caron.js

```diff
@@ -2,270 +2,253 @@
-    '  if jdelay > 0 then',
-    '    local timestamp = (tonumber(ARGV[2]) + jdelay) * 0x1000 + bit.band(jobId, 0xfff)',
-    '    redis.call("ZADD", "bull:" .. jqueue .. ":delayed", timestamp, jobId)',
-    '    redis.call("PUBLISH", "bull:" .. jqueue .. ":delayed", (timestamp / 0x1000))',
-    '  else',
-    '    if redis.call("EXISTS", "bull:" .. jqueue .. ":meta-paused") ~= 1 then',
-    '      redis.call(pushCmd, "bull:" .. jqueue .. ":wait", jobId)',
-    '    else',
-    '      redis.call(pushCmd, "bull:" .. jqueue .. ":paused", jobId)',
-    '    end',
-    '    redis.call("PUBLISH", "bull:" .. jqueue .. ":waiting@null", jobId)',
-    '  end',
-    'end'
-  ].join('\n')
-}
+    this.debug = opts.debug
+    if (this.debug) console.log('started')
 
-scripts.sidekiq = {
-  lua: scripts.ruby_common_1 + '\n' + [
-    'redis.call("LPUSH", "' + program.q_prefix + 'queue:" .. jqueue, payload)'
-  ].join('\n')
-}
+    this.ptype = opts.type
+    this.freq = opts.freq
+    this.kill = opts.kill || true
 
-var script = [scripts.prefix, scripts[ptype].lua, scripts.suffix].join('\n')
+    this.redis = Redis.createClient(opts.redis)
+    this.registerHandlers()
 
-redis.defineCommand('qwork', {
-  lua: script,
-  numberOfKeys: 0
-})
+    const scripts = this.setupScripts(opts)
+    this.setupRedis(scripts)
 
-const elapsedTime = (start) => {
-  return process.hrtime(start)
-}
+    return this
+  }
+
+  stop (cb) {
+    if (!this.status.active) return
 
-const work = () => {
-  if (!STATUS.active || STATUS.processing) return
+    this.status.active = 0
 
-  const rseed = Date.now()
+    if (this.debug) console.log('stopping...')
 
-  if (rseed < STATUS.rseed) {
-    setTimeout(work, 5)
-    return
+    this.stopInter = setInterval(() => {
+      if (this.status.processing) return
+      if (this.debug) console.log('stopped')
+
+      clearInterval(this.stopInter)
+      cb(null)
+    }, 250)
   }
 
-  const args = []
-  var startTs = null
+  start () {
+    this.work()
 
-  if (debug) {
-    startTs = process.hrtime()
+    return this
   }
 
-  args.push(
-    ptype,
-    Date.now(), // timestamp
-    rseed
-  )
+  kill () {
+    if (!this.kill) return
+    process.exit()
+  }
 
-  args.push(
-    (err, res) => {
-      STATUS.processing = 0
+  work () {
+    const debug = this.debug
 
-      if (err) {
-        console.error(err)
-        stop()
-        return
-      }
+    if (!this.status.active || this.status.processing) return
 
-      if (debug) {
-        let elapsed = elapsedTime(startTs)
-        console.log(res[1] + ' jobs processed in ' + elapsed[0] + 's,' + Math.round(elapsed[1] / 1000) + 'µs')
-      }
+    const rseed = Date.now()
 
-      setTimeout(work, program.freq || 25)
+    if (rseed < this.status.rseed) {
+      setTimeout(() => {
+        this.work()
+      }, 5)
```

Comment:
You could set the timeout for `this.status.rseed - rseed` to avoid calling `work()` multiple times